transforms/pair_test.go

```diff
@@ -0,0 +1,244 @@
+		expected:  []optimus.Row{},
+		leftHash:  KeyHasher("k1"),
+		rightHash: KeyHasher("k1"),
+	},
+	{
+		left:  []optimus.Row{a},
+		right: []optimus.Row{a},
+		expected: []optimus.Row{
+			{"left": a, "right": a},
+		},
+		leftHash:  KeyHasher("k1"),
+		rightHash: KeyHasher("k1"),
+	},
+	{
+		left:  []optimus.Row{a},
+		right: []optimus.Row{b},
+		expected: []optimus.Row{
+			{"left": a},
+			{"right": b},
+		},
+		leftHash:  KeyHasher("k1"),
+		rightHash: KeyHasher("k1"),
+	},
+	// Trick test where they hash to nil on both sides
+	{
+		left:  []optimus.Row{c, b},
+		right: []optimus.Row{b, c},
+		expected: []optimus.Row{
+			{"left": c},
+			{"left": b},
+			{"right": b},
+			{"right": c},
+		},
+		leftHash:  KeyHasher("k1"),
+		rightHash: KeyHasher("k1"),
+	},
+	{
+		left:  []optimus.Row{a, b},
+		right: []optimus.Row{a},
+		expected: []optimus.Row{
+			{"left": a, "right": a},
+			{"left": b},
+		},
+		leftHash:  KeyHasher("k1"),
+		rightHash: KeyHasher("k1"),
+	},
+	{
+		left:  []optimus.Row{a},
+		right: []optimus.Row{a, b},
+		expected: []optimus.Row{
+			{"left": a, "right": a},
+			{"right": b},
+		},
+		leftHash:  KeyHasher("k1"),
+		rightHash: KeyHasher("k1"),
+	},
+	{
+		left:  []optimus.Row{a, a},
+		right: []optimus.Row{a, b},
+		expected: []optimus.Row{
+			{"left": a, "right": a},
+			{"left": a, "right": a},
+			{"right": b},
+		},
+		leftHash:  KeyHasher("k1"),
+		rightHash: KeyHasher("k1"),
+	},
+	{
+		left:  []optimus.Row{a, b},
+		right: []optimus.Row{a, a},
+		expected: []optimus.Row{
+			{"left": a, "right": a},
+			{"left": a, "right": a},
+			{"left": b},
+		},
+		leftHash:  KeyHasher("k1"),
+		rightHash: KeyHasher("k1"),
+	},
+	{
+		left:  []optimus.Row{a, a},
+		right: []optimus.Row{a, a},
+		expected: []optimus.Row{
+			{"left": a, "right": a},
+			{"left": a, "right": a},
+			{"left": a, "right": a},
+			{"left": a, "right": a},
+		},
+		leftHash:  KeyHasher("k1"),
+		rightHash: KeyHasher("k1"),
+	},
+	// Now let's get fancy and have one test with different hashes
+	{
+		left:  []optimus.Row{a},
+		right: []optimus.Row{b},
+		expected: []optimus.Row{
+			{"left": a, "right": b},
+		},
+		leftHash:  KeyHasher("k3"),
+		rightHash: KeyHasher("k2"),
+	},
+}
+
+// First, test that we pair correctly by just using an OuterJoin so everything comes through.
+func TestPairSuccess(t *testing.T) {
+	for _, joinTest := range joinTests {
+		left := slice.New(joinTest.left)
+		right := slice.New(joinTest.right)
+
+		pair := Pair(right, joinTest.leftHash, joinTest.rightHash, OuterJoin)
+		table := optimus.Transform(left, pair)
+
+		actual := tests.GetRows(table)
+
+		assert.Equal(t, joinTest.expected, actual)
+		assert.Nil(t, table.Err())
+	}
+}
+
+// Second, test that we filter correctinly on not-outer join by running the same tests with
+// different join types, except filtering the expected array.
+
+var joinFilters = []struct {
+	joinType PairType
+	filter   func(optimus.Row) bool
+}{
+	{
+		joinType: InnerJoin,
```

Comment:
does it make sense to just combine outer join into these tests? we could have a filter that didn't filter anything out? that way we would do all the tests in the same way